# Proyecto Final- TEXT MINING & IMAGE RECOGNITION - SECCIÓN B
# Parte 2
## Diego Obregon
### 11006956

In [2]:
from tensorflow.keras.preprocessing.image import ImageDataGenerator
from tensorflow.keras import layers
from tensorflow.keras import models
from tensorflow.keras.callbacks import EarlyStopping


# Arquitectura 1

In [4]:
def create_model_1():
    model = models.Sequential()
    model.add(layers.Conv2D(32, (3, 3), activation='relu', input_shape=(32, 32, 3)))
    model.add(layers.MaxPooling2D((2, 2)))
    model.add(layers.Conv2D(64, (3, 3), activation='relu'))
    model.add(layers.MaxPooling2D((2, 2)))
    model.add(layers.Flatten())
    model.add(layers.Dense(64, activation='relu'))
    model.add(layers.Dense(6, activation='softmax'))  # 6 clases en total
    
    model.compile(optimizer='adam',
                  loss='categorical_crossentropy',
                  metrics=['accuracy'])
    return model

# Arquitectura 2

In [6]:
def create_model_2():
    model = models.Sequential()
    model.add(layers.Conv2D(32, (3, 3), activation='relu', input_shape=(32, 32, 3)))
    model.add(layers.MaxPooling2D((2, 2)))
    model.add(layers.Conv2D(64, (3, 3), activation='relu'))
    model.add(layers.MaxPooling2D((2, 2)))
    model.add(layers.Conv2D(128, (3, 3), activation='relu'))
    model.add(layers.MaxPooling2D((2, 2)))
    model.add(layers.Flatten())
    model.add(layers.Dense(128, activation='relu'))
    model.add(layers.Dense(6, activation='softmax'))
    
    model.compile(optimizer='adam',
                  loss='categorical_crossentropy',
                  metrics=['accuracy'])
    return model

In [7]:
# Arquitectura 3

In [8]:
def create_model_3():
    model = models.Sequential()
    model.add(layers.Conv2D(32, (3, 3), activation='relu', input_shape=(32, 32, 3)))
    model.add(layers.MaxPooling2D((2, 2)))
    model.add(layers.Conv2D(64, (3, 3), activation='relu'))
    model.add(layers.MaxPooling2D((2, 2)))
    model.add(layers.Conv2D(128, (3, 3), activation='relu'))
    model.add(layers.MaxPooling2D((2, 2)))
    model.add(layers.Flatten())
    model.add(layers.Dense(128, activation='relu'))
    model.add(layers.Dropout(0.5))  # Añadir Dropout para evitar sobreajuste
    model.add(layers.Dense(6, activation='softmax'))
    
    model.compile(optimizer='adam',
                  loss='categorical_crossentropy',
                  metrics=['accuracy'])
    return model


In [9]:
train_datagen = ImageDataGenerator(
    rescale=1./255, 
    rotation_range=40,
    width_shift_range=0.2,
    height_shift_range=0.2,
    shear_range=0.2,
    zoom_range=0.2,
    horizontal_flip=True,
    fill_mode='nearest' )

validation_datagen = ImageDataGenerator(rescale=1./255)

# Conjunto de entrenamiento

In [11]:
train_generator = train_datagen.flow_from_directory(
    'data_Image/train',
    target_size=(32, 32),
    batch_size=32,
    class_mode='categorical')


Found 2572 images belonging to 6 classes.


# Conjunto de validación


In [13]:
validation_generator = validation_datagen.flow_from_directory(
    'data_Image/validation',
    target_size=(32, 32),
    batch_size=32,
    class_mode='categorical')

Found 1106 images belonging to 6 classes.


# CNN

In [15]:
model = models.Sequential()

model.add(layers.Conv2D(32, (3, 3), activation='relu', input_shape=(32, 32, 3)))
model.add(layers.MaxPooling2D((2, 2)))

model.add(layers.Conv2D(64, (3, 3), activation='relu'))
model.add(layers.MaxPooling2D((2, 2)))

model.add(layers.Flatten())
model.add(layers.Dense(128, activation='relu'))
model.add(layers.Dense(6, activation='softmax'))

model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])


C:\Users\USUARIO\anaconda3\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


# Entrenar el modelo

In [17]:
early_stopping = EarlyStopping(monitor='val_loss', patience=3, restore_best_weights=True)

history = model.fit(
    train_generator,
    steps_per_epoch=100,
    epochs=10,
    validation_data=validation_generator,
    validation_steps=50,
    callbacks=[early_stopping] )

Epoch 1/10


C:\Users\USUARIO\anaconda3\Lib\site-packages\keras\src\trainers\data_adapters\py_dataset_adapter.py:122: UserWarning: Your `PyDataset` class should call `super().__init__(**kwargs)` in its constructor. `**kwargs` can include `workers`, `use_multiprocessing`, `max_queue_size`. Do not pass these arguments to `fit()`, as they will be ignored.
  self._warn_if_super_not_called()


 78/100 ━━━━━━━━━━━━━━━━━━━━ 8s 370ms/step - accuracy: 0.3457 - loss: 1.5790 

C:\Users\USUARIO\anaconda3\Lib\contextlib.py:155: UserWarning: Your input ran out of data; interrupting training. Make sure that your dataset or generator can generate at least `steps_per_epoch * epochs` batches. You may need to use the `.repeat()` function when building your dataset.
  self.gen.throw(typ, value, traceback)


100/100 ━━━━━━━━━━━━━━━━━━━━ 47s 419ms/step - accuracy: 0.3681 - loss: 1.5363 - val_accuracy: 0.6383 - val_loss: 0.9560
Epoch 2/10
100/100 ━━━━━━━━━━━━━━━━━━━━ 37s 349ms/step - accuracy: 0.6272 - loss: 0.9743 - val_accuracy: 0.7342 - val_loss: 0.7651
Epoch 3/10
100/100 ━━━━━━━━━━━━━━━━━━━━ 36s 333ms/step - accuracy: 0.6746 - loss: 0.8618 - val_accuracy: 0.6971 - val_loss: 0.8369
Epoch 4/10
100/100 ━━━━━━━━━━━━━━━━━━━━ 34s 305ms/step - accuracy: 0.7234 - loss: 0.7383 - val_accuracy: 0.7703 - val_loss: 0.6660
Epoch 5/10
100/100 ━━━━━━━━━━━━━━━━━━━━ 35s 321ms/step - accuracy: 0.7442 - loss: 0.6771 - val_accuracy: 0.7848 - val_loss: 0.6690
Epoch 6/10
100/100 ━━━━━━━━━━━━━━━━━━━━ 32s 303ms/step - accuracy: 0.7959 - loss: 0.5859 - val_accuracy: 0.6980 - val_loss: 0.8514
Epoch 7/10
100/100 ━━━━━━━━━━━━━━━━━━━━ 33s 314ms/step - accuracy: 0.7871 - loss: 0.5844 - val_accuracy: 0.8662 - val_loss: 0.4146
Epoch 8/10
100/100 ━━━━━━━━━━━━━━━━━━━━ 34s 332ms/step - accuracy: 0.8092 - loss: 0.5024 - val

# Entrenar y evaluar Modelo 1

In [19]:
model_1 = create_model_1()
history_1 = model_1.fit(train_generator, epochs=10, validation_data=validation_generator, callbacks=[early_stopping])
test_loss_1, test_acc_1 = model_1.evaluate(validation_generator)
print(f'Model 1 Accuracy: {test_acc_1}')

C:\Users\USUARIO\anaconda3\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/10
81/81 ━━━━━━━━━━━━━━━━━━━━ 43s 460ms/step - accuracy: 0.2893 - loss: 1.6614 - val_accuracy: 0.4286 - val_loss: 1.4363
Epoch 2/10
81/81 ━━━━━━━━━━━━━━━━━━━━ 38s 432ms/step - accuracy: 0.5309 - loss: 1.1941 - val_accuracy: 0.6257 - val_loss: 1.0534
Epoch 3/10
81/81 ━━━━━━━━━━━━━━━━━━━━ 40s 467ms/step - accuracy: 0.6351 - loss: 0.9534 - val_accuracy: 0.7315 - val_loss: 0.7794
35/35 ━━━━━━━━━━━━━━━━━━━━ 9s 270ms/step - accuracy: 0.4327 - loss: 1.4195
Model 1 Accuracy: 0.4285714328289032


In [20]:
# Entrenar y evaluar Modelo 2

In [21]:
model_2 = create_model_2()
history_2 = model_2.fit(train_generator, epochs=10, validation_data=validation_generator, callbacks=[early_stopping])
test_loss_2, test_acc_2 = model_2.evaluate(validation_generator)
print(f'Model 2 Accuracy: {test_acc_2}')

Epoch 1/10
81/81 ━━━━━━━━━━━━━━━━━━━━ 76s 806ms/step - accuracy: 0.2967 - loss: 1.6334 - val_accuracy: 0.5262 - val_loss: 1.1500
Epoch 2/10
81/81 ━━━━━━━━━━━━━━━━━━━━ 54s 603ms/step - accuracy: 0.5831 - loss: 1.0690 - val_accuracy: 0.6212 - val_loss: 0.9162
Epoch 3/10
81/81 ━━━━━━━━━━━━━━━━━━━━ 43s 515ms/step - accuracy: 0.6605 - loss: 0.8792 - val_accuracy: 0.7405 - val_loss: 0.7170
35/35 ━━━━━━━━━━━━━━━━━━━━ 11s 314ms/step - accuracy: 0.5252 - loss: 1.1776
Model 2 Accuracy: 0.5262206196784973


In [22]:
# Entrenar y evaluar Modelo 3

In [23]:
model_3 = create_model_3()
history_3 = model_3.fit(train_generator, epochs=10, validation_data=validation_generator, callbacks=[early_stopping])
test_loss_3, test_acc_3 = model_3.evaluate(validation_generator)
print(f'Model 3 Accuracy: {test_acc_3}')

Epoch 1/10
81/81 ━━━━━━━━━━━━━━━━━━━━ 43s 467ms/step - accuracy: 0.2377 - loss: 1.7111 - val_accuracy: 0.4819 - val_loss: 1.2795
Epoch 2/10
81/81 ━━━━━━━━━━━━━━━━━━━━ 43s 488ms/step - accuracy: 0.4797 - loss: 1.3109 - val_accuracy: 0.5958 - val_loss: 0.9774
Epoch 3/10
81/81 ━━━━━━━━━━━━━━━━━━━━ 37s 420ms/step - accuracy: 0.6157 - loss: 1.0126 - val_accuracy: 0.6881 - val_loss: 0.7848
35/35 ━━━━━━━━━━━━━━━━━━━━ 9s 247ms/step - accuracy: 0.4754 - loss: 1.2770
Model 3 Accuracy: 0.4819168150424957


# Evaluacion del modelo



In [25]:
test_loss, test_acc = model.evaluate(validation_generator, steps=50)
print('Test accuracy:', test_acc)


50/50 ━━━━━━━━━━━━━━━━━━━━ 9s 175ms/step - accuracy: 0.8761 - loss: 0.4013
Test accuracy: 0.876130223274231


In [26]:
print(f'Model 1 Accuracy: {test_acc_1}')
print(f'Model 2 Accuracy: {test_acc_2}')
print(f'Model 3 Accuracy: {test_acc_3}')


Model 1 Accuracy: 0.4285714328289032
Model 2 Accuracy: 0.5262206196784973
Model 3 Accuracy: 0.4819168150424957


# Conclucion

### Modelo 1, la arquitectura de este modelo es basicacon dos capas convolucionales es insuficiente para capturar los patrones necesarios y se obtinen la precisión más baja (42.9%).

### Modelo 2, este modelo cuenta con tres capas convolucionales y mejora el rendimiento, alcanzando la mejor precisión (52.6%)

### Modelo 3, este modeo incluye dropout para evitar los sobreajustes, aun asi su rendimiento es inferior al del Modelo 2 (48.2%)